test/model/test/model_transaction/main.cpp

```diff
@@ -352,6 +365,308 @@ test_transaction_basic_wt(void)
+    wt_model_txn_begin_both(txn1, session1, 10);
+    wt_model_txn_assert(table, uri, txn1, session1, recno1);
+    wt_model_txn_commit_both(txn1, session1, 15);
+
+    /* Check transaction conflicts: Concurrent update. */
+    wt_model_txn_begin_both(txn1, session1);
+    wt_model_txn_begin_both(txn2, session2);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value3);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value4);
+    wt_model_txn_insert_both(table, uri, txn2, session2, recno1, value4); /* Rollback. */
+    wt_model_txn_commit_both(txn1, session1, 20);
+    wt_model_txn_commit_both(txn2, session2, 20);
+
+    wt_model_txn_begin_both(txn1, session1);
+    wt_model_txn_begin_both(txn2, session2);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value3);
+    wt_model_txn_commit_both(txn1, session1, 30);
+    wt_model_txn_insert_both(table, uri, txn2, session2, recno1, value4); /* Rollback. */
+    wt_model_txn_commit_both(txn2, session2, 30);
+
+    /* Check transaction conflicts: Update not in the transaction snapshot. */
+    wt_model_txn_begin_both(txn1, session1);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value1);
+    wt_model_txn_begin_both(txn2, session2);
+    wt_model_txn_commit_both(txn1, session1, 40);
+    wt_model_txn_insert_both(table, uri, txn2, session2, recno1, value4); /* Rollback. */
+    wt_model_txn_commit_both(txn2, session2, 40);
+
+    // Not testing conflict between a transactional update and an update outside of a transaction;
+    // this can result in a hang or an abort.
+
+    /* Set timestamp. */
+    wt_model_txn_begin_both(txn1, session1);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value4);
+    wt_model_txn_set_timestamp_both(txn1, session1, 43);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno2, value5);
+    wt_model_txn_set_timestamp_both(txn1, session1, 44);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno3, value6);
+    wt_model_txn_commit_both(txn1, session1, 50);
+
+    wt_model_assert(table, uri, recno1, 50);
+    wt_model_assert(table, uri, recno2, 42);
+    wt_model_assert(table, uri, recno3, 44);
+    wt_model_assert(table, uri, recno1, 50 - 1);
+    wt_model_assert(table, uri, recno2, 42 - 1);
+    wt_model_assert(table, uri, recno3, 44 - 1);
+
+    /* Set timestamp: Check timestamp order within the same recno. */
+    wt_model_txn_begin_both(txn1, session1);
+    wt_model_txn_set_timestamp_both(txn1, session1, 52);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value1);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value2);
+    wt_model_txn_set_timestamp_both(txn1, session1, 55);
+    wt_model_txn_insert_both(table, uri, txn1, session1, recno1, value3);
+    wt_model_txn_set_timestamp_both(txn1, session1, 53);
+    // Cannot insert recno 1 at timestamp 53: Commit would result in abort.
```

Comment:
Abort or conflict?

Reply:
Yes, this one results in an abort:
```
[1714065658:944114][2358800:0x7f0811f81b80], WT_SESSION.commit_transaction: [WT_VERB_DEFAULT][ERROR]: int __txn_timestamp_usage_check(WT_SESSION_IMPL *, WT_TXN_OP *, WT_UPDATE *), 1003: file:table.wt: unexpected timestamp usage: updating a value with a timestamp (0, 53) before the previous update (0, 55): Invalid argument
[1714065658:944150][2358800:0x7f0811f81b80], WT_SESSION.commit_transaction: [WT_VERB_DEFAULT][ERROR]: void __wt_abort(WT_SESSION_IMPL *), 28: aborting WiredTiger library
Aborted
```

Reply:
Thank you. Any reason why we keep this as a comment?

Reply:
The model emulates this behavior, which is tested in a non-recno, non-WT version of the test. That test is a complete parallel of this test, except that it can't check this behavior, so the corresponding part is still there but commented out. It also gives me a code to uncomment and test if I want to check whether the behavior still results in an abort.

Reply:
Got it!